File: 3492-count-submatrices-with-equal-frequency-of-x-and-y/count-submatrices-with-equal-frequency-of-x-and-y.java
class Solution {
    public int numberOfSubmatrices(char[][] grid) {
        int m = grid.length, n = grid[0].length;
        int[][] pfX = new int[m + 1][n + 1];
        int[][] pfY = new int[m + 1][n + 1];
        for (int i = 1; i <= m; i++) {
            for (int j = 1; j <= n; j++) {
                pfX[i][j] = pfX[i - 1][j] + pfX[i][j - 1] - pfX[i - 1][j - 1];
                pfY[i][j] = pfY[i - 1][j] + pfY[i][j - 1] - pfY[i - 1][j - 1];
                if (grid[i - 1][j - 1] == 'X') pfX[i][j]++;
                if (grid[i - 1][j - 1] == 'Y') pfY[i][j]++;
            }
        }
        int count = 0;
        for (int i2 = 1; i2 <= m; i2++) {
            for (int j2 = 1; j2 <= n; j2++) {
                int x = pfX[i2][j2];
                int y = pfY[i2][j2];
                if (x == y && x > 0) count++;
            }
        }
        return count;
    }
}

File: 3492-count-submatrices-with-equal-frequency-of-x-and-y/count-submatrices-with-equal-frequency-of-x-and-y.rs
impl Solution {
    pub fn number_of_submatrices(grid: Vec<Vec<char>>) -> i32 {
        let m = grid.len();
        let n = grid[0].len();
        let mut pfx = vec![vec![0; n + 1]; m + 1];
        let mut pfy = vec![vec![0; n + 1]; m + 1];
        for i in 1..=m {
            for j in 1..=n {
                pfx[i][j] = pfx[i - 1][j] + pfx[i][j - 1] - pfx[i - 1][j - 1];
                pfy[i][j] = pfy[i - 1][j] + pfy[i][j - 1] - pfy[i - 1][j - 1];
                if grid[i - 1][j - 1] == 'X' {
                    pfx[i][j] += 1;
                }
                if grid[i - 1][j - 1] == 'Y' {
                    pfy[i][j] += 1;
                }
            }
        }
        let mut count = 0;
        for i2 in 1..=m {
            for j2 in 1..=n {
                let x = pfx[i2][j2];
                let y = pfy[i2][j2];
                if x == y && x > 0 {
                    count += 1;
                }
            }
        }
        count
    }
}